Sources/Workspace/LocalContainerProvider.swift

```diff
@@ -0,0 +1,111 @@
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageModel
+import PackageGraph
+import TSCBasic
+import TSCUtility
+import SourceControl
+
+/// PackageContainerProvider implementation used by Workspace to do a dependency pre-calculation using the cached
+/// dependency information (Workspace.DependencyManifests) to check if dependency resolution is required before
+/// performing a full resolution.
+struct LocalContainerProvider: PackageContainerProvider {
+    let root: PackageGraphRoot
+    let currentManifests: Workspace.DependencyManifests
+    let config: SwiftPMConfig
+    let diagnostics: DiagnosticsEngine
+
+    func getContainer(
+        for identifier: PackageReference,
+        skipUpdate: Bool,
+        completion: @escaping (Result<PackageContainer, AnyError>) -> Void
+    ) {
+        if let manifest = currentManifests.dependencies.first(where: { $1.packageRef == identifier }) {
+            let container = LocalPackageContainer(
+                manifest: manifest.manifest,
+                dependency: manifest.dependency,
+                config: config
+            )
+
+            return completion(.success(container))
+        }
+
+        if let index = currentManifests.root.packageRefs.firstIndex(of: identifier) {
+            let container = LocalPackageContainer(
+                manifest: currentManifests.root.manifests[index],
+                dependency: nil,
+                config: config
+            )
+
+            return completion(.success(container))
+        }
+
+        completion(.init(Diagnostics.fatalError))
+    }
+}
+
+private struct LocalPackageContainer: PackageContainer {
+    let manifest: Manifest
+    let dependency: ManagedDependency?
+    let config: SwiftPMConfig
+
+    var identifier: PackageReference {
+        if let identifier = dependency?.packageRef {
+            return identifier
+        } else {
+            let identity = PackageReference.computeIdentity(packageURL: manifest.url)
+            return PackageReference(
+                identity: identity,
+                path: manifest.path.pathString,
+                isLocal: true
+            )
+        }
+    }
+
+    var reversedVersions: [Version] {
+        guard let dependency = self.dependency else { return [] }
+
+        switch dependency.state {
+        case .checkout(let state):
+            if let version = state.version {
+                return [version]
+            }
+        case .edited, .local:
+            break
+        }
+
+        return []
+    }
+
+    func isToolsVersionCompatible(at version: Version) -> Bool {
+        return true
+    }
+
+    func versions(filter isIncluded: (Version) -> Bool) -> AnySequence<Version> {
+        return AnySequence(reversedVersions)
+    }
+
+    func getDependencies(at version: Version) throws -> [PackageContainerConstraint] {
+        return manifest.dependencyConstraints(config: config)
```

Comment:
we should add an assert that dependency is at this version

Reply:
Done.

Reply:
Same comment as above, but the other way round.